controllers/sandbox_controller.go

```diff
@@ -0,0 +1,155 @@
+	log := log.FromContext(ctx)
+
+	sandbox := &sandboxv1alpha1.Sandbox{}
+	if err := r.Get(ctx, req.NamespacedName, sandbox); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("sandbox resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get sandbox")
+		return ctrl.Result{}, err
+	}
+
+	// Check if the sandbox is being deleted
+	if !sandbox.ObjectMeta.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(sandbox, sandboxFinalizer) {
+			if err := r.finalizeSandbox(ctx, sandbox); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			controllerutil.RemoveFinalizer(sandbox, sandboxFinalizer)
+			if err := r.Update(ctx, sandbox); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Add finalizer for this CR
+	if !controllerutil.ContainsFinalizer(sandbox, sandboxFinalizer) {
+		controllerutil.AddFinalizer(sandbox, sandboxFinalizer)
+		if err := r.Update(ctx, sandbox); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Check if the pod already exists, if not create a new one
+	found := &corev1.Pod{}
+	err := r.Get(ctx, types.NamespacedName{Name: sandbox.Name, Namespace: sandbox.Namespace}, found)
+	if err != nil && errors.IsNotFound(err) {
+		// Define a new pod
+		pod, err := r.podForSandbox(sandbox)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		log.Info("Creating a new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
+		if err = r.Create(ctx, pod); err != nil {
+			log.Error(err, "Failed to create new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	} else if err != nil {
+		log.Error(err, "Failed to get Pod")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *SandboxReconciler) finalizeSandbox(ctx context.Context, s *sandboxv1alpha1.Sandbox) error {
+	pod := &corev1.Pod{}
+	err := r.Get(ctx, types.NamespacedName{Name: s.Name, Namespace: s.Namespace}, pod)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get pod during finalization: %w", err)
+	}
+
+	if err := r.Delete(ctx, pod); err != nil {
+		return fmt.Errorf("failed to delete pod during finalization: %w", err)
+	}
+	return nil
+}
+
+func (r *SandboxReconciler) podForSandbox(s *sandboxv1alpha1.Sandbox) (*corev1.Pod, error) {
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+		},
+		Spec: s.Spec.Template.Spec,
+	}
+	if err := ctrl.SetControllerReference(s, pod, r.Scheme); err != nil {
+		return nil, err
+	}
+	return pod, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SandboxReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&sandboxv1alpha1.Sandbox{}).
+		Owns(&corev1.Pod{}).
```

Comment:
I _think_ this sets up a watch (?) and this is what I think would be great to filter by label, because pod watches are not cheap

Reply:
added label predicate